File: hdf-diff/src/summary.ts
import type { ComparisonSummary, RequirementDiff } from './types.js';

/**
 * Compute summary counts from an array of RequirementDiff entries.
 */
export function computeSummary(requirements: RequirementDiff[]): ComparisonSummary {
  const summary: ComparisonSummary = {
    fixed: 0,
    regressed: 0,
    new: 0,
    absent: 0,
    unchanged: 0,
    updated: 0,
    total: requirements.length,
    matchedCount: 0,
    unmatchedOldCount: 0,
    unmatchedNewCount: 0,
  };

  for (const req of requirements) {
    switch (req.state) {
      case 'fixed':
        summary.fixed++;
        summary.matchedCount++;
        break;
      case 'regressed':
        summary.regressed++;
        summary.matchedCount++;
        break;
      case 'new':
        summary.new++;
        summary.unmatchedNewCount++;
        break;
      case 'absent':
        summary.absent++;
        summary.unmatchedOldCount++;
        break;
      case 'unchanged':
        summary.unchanged++;
        summary.matchedCount++;
        break;
      case 'updated':
        summary.updated++;
        summary.matchedCount++;
        break;
    }
  }

  return summary;
}

File: hdf-diff/go/summary.go
package diff

// ComputeSummary computes aggregate counts from requirement diffs.
func ComputeSummary(requirements []RequirementDiff) ComparisonSummary {
	summary := ComparisonSummary{Total: len(requirements)}

	for _, req := range requirements {
		switch req.State {
		case StateFixed:
			summary.Fixed++
			summary.MatchedCount++
		case StateRegressed:
			summary.Regressed++
			summary.MatchedCount++
		case StateNew:
			summary.New++
			summary.UnmatchedNewCount++
		case StateAbsent:
			summary.Absent++
			summary.UnmatchedOldCount++
		case StateUnchanged:
			summary.Unchanged++
			summary.MatchedCount++
		case StateUpdated:
			summary.Updated++
			summary.MatchedCount++
		case StateMoved, StateSplit, StateMerged:
			// Reserved for v1.1 — no-op in summary counting.
			summary.MatchedCount++
		}
	}

	return summary
}
